File: 002.Artificial_Intelligence/012.AIAgent/002.tool/LangChain/day_03_langchain_tool/app/infrastructure/persistence/chat_store.py
# -*- coding: utf-8 -*-
"""SQLite 会话/消息仓储。Demo 固定 user_id=1。"""
from __future__ import annotations

import sqlite3
from pathlib import Path

from app.domain.model import ChatMessage, ChatSession

FIXED_USER_ID = 1


class ChatStore:
    def __init__(self, db_path: Path) -> None:
        self.db_path = db_path
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._init_schema()

    def _connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self.db_path)
        conn.row_factory = sqlite3.Row
        return conn

    def _init_schema(self) -> None:
        with self._connect() as conn:
            conn.executescript(
                """
                CREATE TABLE IF NOT EXISTS chat_session (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    user_id INTEGER NOT NULL,
                    title TEXT NOT NULL,
                    created_at TEXT NOT NULL DEFAULT (datetime('now','localtime'))
                );
                CREATE TABLE IF NOT EXISTS chat_message (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    session_id INTEGER NOT NULL,
                    role TEXT NOT NULL,
                    content TEXT NOT NULL,
                    created_at TEXT NOT NULL DEFAULT (datetime('now','localtime')),
                    FOREIGN KEY(session_id) REFERENCES chat_session(id)
                );
                """
            )

    def create_session(self, title: str = "新对话", user_id: int = FIXED_USER_ID) -> ChatSession:
        with self._connect() as conn:
            cur = conn.execute(
                "INSERT INTO chat_session(user_id, title) VALUES (?, ?)",
                (user_id, title),
            )
            sid = int(cur.lastrowid)
            row = conn.execute("SELECT * FROM chat_session WHERE id = ?", (sid,)).fetchone()
        return self._session(row)

    def list_sessions(self, user_id: int = FIXED_USER_ID) -> list[ChatSession]:
        with self._connect() as conn:
            rows = conn.execute(
                "SELECT * FROM chat_session WHERE user_id = ? ORDER BY id DESC",
                (user_id,),
            ).fetchall()
        return [self._session(r) for r in rows]

    def get_session(self, session_id: int, user_id: int = FIXED_USER_ID) -> ChatSession | None:
        with self._connect() as conn:
            row = conn.execute(
                "SELECT * FROM chat_session WHERE id = ? AND user_id = ?",
                (session_id, user_id),
            ).fetchone()
        return self._session(row) if row else None

    def list_messages(self, session_id: int) -> list[ChatMessage]:
        with self._connect() as conn:
            rows = conn.execute(
                "SELECT * FROM chat_message WHERE session_id = ? ORDER BY id ASC",
                (session_id,),
            ).fetchall()
        return [self._message(r) for r in rows]

    def add_message(self, session_id: int, role: str, content: str) -> ChatMessage:
        with self._connect() as conn:
            cur = conn.execute(
                "INSERT INTO chat_message(session_id, role, content) VALUES (?, ?, ?)",
                (session_id, role, content),
            )
            mid = int(cur.lastrowid)
            # ponytail: 首条用户消息时用前 20 字更新标题，避免全叫「新对话」
            if role == "user":
                title_row = conn.execute(
                    "SELECT COUNT(*) AS c FROM chat_message WHERE session_id = ?",
                    (session_id,),
                ).fetchone()
                if int(title_row["c"]) == 1:
                    conn.execute(
                        "UPDATE chat_session SET title = ? WHERE id = ?",
                        (content.strip()[:20] or "新对话", session_id),
                    )
            row = conn.execute("SELECT * FROM chat_message WHERE id = ?", (mid,)).fetchone()
        return self._message(row)

    @staticmethod
    def _session(row: sqlite3.Row) -> ChatSession:
        return ChatSession(
            id=int(row["id"]),
            user_id=int(row["user_id"]),
            title=str(row["title"]),
            created_at=str(row["created_at"]),
        )

    @staticmethod
    def _message(row: sqlite3.Row) -> ChatMessage:
        return ChatMessage(
            id=int(row["id"]),
            session_id=int(row["session_id"]),
            role=str(row["role"]),
            content=str(row["content"]),
            created_at=str(row["created_at"]),
        )

File: 002.Artificial_Intelligence/012.AIAgent/000.准备阶段/002.聊天管理/Spring AI/demo_002_spring_ai_chat_session/src/main/java/com/demo/springai/session/infrastructure/persistence/ChatStore.java
package com.demo.springai.session.infrastructure.persistence;

import com.demo.springai.session.domain.model.ChatMessage;
import com.demo.springai.session.domain.model.ChatSession;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;
import org.springframework.stereotype.Repository;

import java.sql.PreparedStatement;
import java.sql.Statement;
import java.util.List;
import java.util.Optional;

/**
 * SQLite 会话/消息仓储。Demo 固定 user_id=1。
 */
@Repository
public class ChatStore {

    public static final long FIXED_USER_ID = 1L;

    private final JdbcTemplate jdbc;
    private final long fixedUserId;

    private static final RowMapper<ChatSession> SESSION_MAPPER = (rs, i) -> new ChatSession(
            rs.getLong("id"),
            rs.getLong("user_id"),
            rs.getString("title"),
            rs.getString("created_at")
    );

    private static final RowMapper<ChatMessage> MESSAGE_MAPPER = (rs, i) -> new ChatMessage(
            rs.getLong("id"),
            rs.getLong("session_id"),
            rs.getString("role"),
            rs.getString("content"),
            rs.getString("created_at")
    );

    public ChatStore(
            JdbcTemplate jdbc,
            @Value("${demo.chat.fixed-user-id:1}") long fixedUserId) {
        this.jdbc = jdbc;
        this.fixedUserId = fixedUserId;
    }

    @PostConstruct
    public void initSchema() {
        jdbc.execute("""
                CREATE TABLE IF NOT EXISTS chat_session (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    user_id INTEGER NOT NULL,
                    title TEXT NOT NULL,
                    created_at TEXT NOT NULL DEFAULT (datetime('now','localtime'))
                )
                """);
        jdbc.execute("""
                CREATE TABLE IF NOT EXISTS chat_message (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    session_id INTEGER NOT NULL,
                    role TEXT NOT NULL,
                    content TEXT NOT NULL,
                    created_at TEXT NOT NULL DEFAULT (datetime('now','localtime')),
                    FOREIGN KEY(session_id) REFERENCES chat_session(id)
                )
                """);
    }

    public ChatSession createSession(String title) {
        KeyHolder keyHolder = new GeneratedKeyHolder();
        jdbc.update(con -> {
            PreparedStatement ps = con.prepareStatement(
                    "INSERT INTO chat_session(user_id, title) VALUES (?, ?)",
                    Statement.RETURN_GENERATED_KEYS
            );
            ps.setLong(1, fixedUserId);
            ps.setString(2, title == null || title.isBlank() ? "新对话" : title.trim());
            return ps;
        }, keyHolder);
        long id = requireGeneratedId(keyHolder);
        return getSession(id).orElseThrow();
    }

    public List<ChatSession> listSessions() {
        return jdbc.query(
                "SELECT * FROM chat_session WHERE user_id = ? ORDER BY id DESC",
                SESSION_MAPPER,
                fixedUserId
        );
    }

    public Optional<ChatSession> getSession(long sessionId) {
        List<ChatSession> list = jdbc.query(
                "SELECT * FROM chat_session WHERE id = ? AND user_id = ?",
                SESSION_MAPPER,
                sessionId,
                fixedUserId
        );
        return list.stream().findFirst();
    }

    public List<ChatMessage> listMessages(long sessionId) {
        return jdbc.query(
                "SELECT * FROM chat_message WHERE session_id = ? ORDER BY id ASC",
                MESSAGE_MAPPER,
                sessionId
        );
    }

    public ChatMessage addMessage(long sessionId, String role, String content) {
        KeyHolder keyHolder = new GeneratedKeyHolder();
        jdbc.update(con -> {
            PreparedStatement ps = con.prepareStatement(
                    "INSERT INTO chat_message(session_id, role, content) VALUES (?, ?, ?)",
                    Statement.RETURN_GENERATED_KEYS
            );
            ps.setLong(1, sessionId);
            ps.setString(2, role);
            ps.setString(3, content);
            return ps;
        }, keyHolder);

        if ("user".equals(role)) {
            Integer count = jdbc.queryForObject(
                    "SELECT COUNT(*) FROM chat_message WHERE session_id = ?",
                    Integer.class,
                    sessionId
            );
            if (count != null && count == 1) {
                String title = content == null ? "新对话" : content.strip();
                if (title.length() > 20) {
                    title = title.substring(0, 20);
                }
                if (title.isBlank()) {
                    title = "新对话";
                }
                jdbc.update("UPDATE chat_session SET title = ? WHERE id = ?", title, sessionId);
            }
        }

        long mid = requireGeneratedId(keyHolder);
        return jdbc.query(
                "SELECT * FROM chat_message WHERE id = ?",
                MESSAGE_MAPPER,
                mid
        ).get(0);
    }

    private static long requireGeneratedId(KeyHolder keyHolder) {
        Number key = keyHolder.getKey();
        if (key == null) {
            throw new IllegalStateException("未能获取自增主键");
        }
        return key.longValue();
    }
}
